src/ProfileEvent.h

```diff
@@ -0,0 +1,358 @@
+    }
+
+    /** Print event */
+    void print(std::ostream& os) override {
+        os << ProfileKeySingleton::instance().getText(getKey());
+        /* system CPU time used */
+        double t = (double)ru.ru_stime.tv_sec * 1000000.0 + (double)ru.ru_stime.tv_usec;
+        os << getStartString() << ",";
+        os << std::to_string(t);
+        // os << "," << getCurrentValue();
+        os << std::endl;
+    }
+};
+
+/**
+ * Profile Memory Event
+ */
+class ProfileMemoryEvent : public ProfileEvent {
+private:
+    /** resource statistics */
+    struct rusage ru;
+
+public:
+    ProfileMemoryEvent(const std::string& txt) : ProfileEvent(txt) {
+        getrusage(RUSAGE_SELF, &ru);
+    }
+
+    /** Print event */
+    void print(std::ostream& os) override {
+        os << ProfileKeySingleton::instance().getText(getKey());
+        os << getStartString() << ",";
+        os << std::to_string(ru.ru_maxrss); /* maximum resident set size */
+        os << std::endl;
+    }
+};
+
+/**
+ * Profile Event Singleton
+ */
+class ProfileEventSingleton {
+private:
+    /**  Profile Timer */
+    class ProfileTimer {
+    private:
+        /** time interval between per utilisation read */
+        Interval t;
+
+        /** timer is running */
+        bool running;
+
+        /** thread timer runs on */
+        std::thread th;
+
+        /** run method for thread th */
+        void run() {
+            ProfileEventSingleton::instance().makeUtilisationEvent("utilisation");
+            ProfileEventSingleton::instance().makeMemoryEvent("memory");
+        }
+
+        Interval getInterval() {
+            return t;
+        }
+
+        bool getRunning() {
+            return running;
+        }
+
+    public:
+        ProfileTimer(Interval in = 1) : t(in), running(false) {}
+
+        /** start timer on the thread th */
+        void start() {
+            running = true;
+
+            th = std::thread([this]() {
+                while (this->getRunning()) {
+                    this->run();
+                    auto x =
+                            std::chrono::steady_clock::now() + std::chrono::milliseconds(this->getInterval());
+                    std::this_thread::sleep_until(x);
+                }
+            });
+        }
+
+        /** stop timer on the thread th */
+        void stop() {
+            running = false;
+            th.join();
+        }
+    };
+
+private:
+    ProfileEventSingleton() = default;
+    std::list<ProfileEvent*> events;
+    std::mutex eventMutex;
+    ProfileTimer timer;
+
+public:
+    ~ProfileEventSingleton() {
+        for (auto const& cur : events) {
+            delete cur;
+        }
+    }
+    static ProfileEventSingleton& instance() {
+        static ProfileEventSingleton singleton;
+        return singleton;
+    }
+
+    /** Make new time event */
+    ProfileTimingEvent* makeTimingEvent(const std::string& txt) {
+        ProfileTimingEvent* e = new ProfileTimingEvent(txt);
```

Comment:
Save a little time for the reader and use `auto`, since you name the class twice on one line.